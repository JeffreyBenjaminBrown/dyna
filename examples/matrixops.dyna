% A and B are names of matrices
times(A, B, I, J) += m(A, I, K) * m(B, K, J) for _ is product(A,B).

m(P, I, J) += shape(A,R,C), % pair(R, C) is shape(A),
              shape(B,C,_), % pair(C, D) is shape(B),
              P is product(A, B),
              times(A, B, I, J).

%shape(P, R, C) :- P is product(A, B), shape(A, R, X), shape(B, X, C).
%shape(A) += &pair(rows(A), cols(A)).

shape(X, rows(X), cols(X)).
rows(X) max= _ is m(X, R, _), R.
cols(X) max= _ is m(X, _, C), C.

% define some matrices
:-dispos m(&,*,*).

% matrix "a" = [ 1 0 ;
%                0 1 ]
m(a, 1, 1) += 1.
m(a, 1, 2) += 0.
m(a, 2, 1) += 0.
m(a, 2, 2) += 1.

% matrix "b" = [ 3 0 1 ;
%                0 2 0 ]
m(b, 1, 1) += 3.
m(b, 1, 2) += 0.
m(b, 1, 3) += 1.
m(b, 2, 1) += 0.
m(b, 2, 2) += 2.
m(b, 2, 3) += 0.

% matrix "c" is the product of matricies "a" and "b"
:-dispos product(&,&).
product(a,b) := &c.
