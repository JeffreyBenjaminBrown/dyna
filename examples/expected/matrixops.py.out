
Solution
========

cols/1
======
cols(a) = 2.
cols(b) = 3.
cols(c) = 3.

m/3
===
m(a,1,1) = 1.
m(a,1,2) = 0.
m(a,2,1) = 0.
m(a,2,2) = 1.
m(b,1,1) = 3.
m(b,1,2) = 0.
m(b,1,3) = 1.
m(b,2,1) = 0.
m(b,2,2) = 2.
m(b,2,3) = 0.
m(c,1,1) = 3.
m(c,1,2) = 0.
m(c,1,3) = 1.
m(c,2,1) = 0.
m(c,2,2) = 2.
m(c,2,3) = 0.

product/2
=========
product(a,b) = c.

rows/1
======
rows(a) = 2.
rows(b) = 2.
rows(c) = 2.

shape/3
=======
shape(a,2,2).
shape(b,2,3).
shape(c,2,3).

times/4
=======
times(a,b,1,1) = 3.
times(a,b,1,2) = 0.
times(a,b,1,3) = 1.
times(a,b,2,1) = 0.
times(a,b,2,2) = 2.
times(a,b,2,3) = 0.

